### DISCUSSION TITLE
To Soft Delete or Not to Soft Delete

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#114] [General]
# To Soft Delete or Not to Soft Delete

In this blog post I will discuss the pros and cons of soft deleting records in a database versus other approaches you can take. Heat up your pop corn and get your salt ready!

https://steven-giesel.com/blogPost/a807373c-dcc6-42f9-995f-e69dcea1cd47

**Comment by diegodrf** [comment]
Thanks for the post Steven.

You showed an alternative to move the obsolete data to a new table/database, but you didn't cover how to do it.

I can imagine two ways to implement it.

First, I do it at the transaction time, so I should to insert the new data, but I should to read the old data, insert into the new table, remove the old data and complete my commit.
Second, I create a column to check the data as obsolete (so here we return to the start point creating an IsDeleted column) and a background task will clean up my database searching for obsolete values to move to the Archive table.
The first one has performance issues too, we just remove the problem from the read step to the write step. And the second one keep the IsDeleted column to allows the background task to track the rows.

Do you have a better alternative for it?

### Alternative
If we are talking about SQL Server,  [temporal tables](https://learn.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver16) can be an alternative for it. I confess that I don't know too much about it, but I worked on a system that used it. But it won't eliminate the database size problem.

**Reply by linkdotnet** [reply]
Thanks Diego,

I deliberately left this out, as there are many approaches to do so.
Often times, you have a `HttpDelete` endpoint that is used. Here you can easily move the data from one table/schema to another - or move it to another database. Obviously the challenge is here to keep both db's in a consistent state.

@diegodrf you mentioned you might need this while you are adding data:
> First, I do it at the transaction time, so I should to insert the new data, but I should to read the old data, insert into the new table, remove the old data and complete my commit.

Did I grasp that correctly?

**Reply by diegodrf** [reply]
I thought a query like this to describe this step.

>First, I do it at the transaction time, so I should to insert the new data, but I should to read the old data, insert into the new table, remove the old data and complete my commit.

```sql
BEGIN TRANSACTION;

-- I should to insert the new data
INSERT INTO [OriginalTable] (...) VALUES (...);

-- but I should to read the old data, insert into the new table
INSERT INTO [ArchiveTable] 
SELECT * FROM [OriginalTable] 
WHERE IsDeleted = 1;

-- remove the old data
DELETE FROM [OriginalTable] WHERE IsDeleted = 1;

-- and complete my commit
COMMIT;
```

**Reply by diegodrf** [reply]
Sorry, reading again I realize that I described it as a new register. 

Let me change it for an update. 

But the scenario is the same. It should run inside a transaction degrading the write performance.

```sql
BEGIN TRANSACTION;

INSERT INTO [ArchiveTable] 
SELECT * FROM [OriginalTable] 
WHERE Id = 123;

UPDATE [OriginalTable] SET ... WHERE Id = 123;

COMMIT;
```

**Reply by linkdotnet** [reply]
Now I understand - yes. 

I would still try to avoid the `IsDeleted` flag because now you have the handling plus the complexity of the background service.
Of course you gain some space/performance by regularly cleaning up.
